packages/my-queries-storage/src/provider.ts

```diff
@@ -0,0 +1,51 @@
+import { createContext, useContext } from 'react';
+import type { PipelineStorage } from './pipeline-storage';
+import type {
+  FavoriteQueryStorage,
+  RecentQueryStorage,
+  QueryStorageOptions,
+} from './query-storage';
+
+export type FavoriteQueryStorageAccess = {
+  createStorage(options?: QueryStorageOptions): FavoriteQueryStorage;
+};
+
+export type RecentQueryStorageAccess = {
+  createStorage(options?: QueryStorageOptions): RecentQueryStorage;
```

Comment:
Two things:
- This is just a really tiny thing, but it should be transparent to the caller whether `createStorage()` genuinely returns a new object or not, it never actually creates a resource, so should these be called `getStorage()` to make that a bit clearer?
- Right now the return types for these methods are class types, that means that this is not super usable as an _interface_ (that e.g. Data Explorer could implement differently), would it be possible to turn this into one? I know that that adds a bit of complexity, but otherwise dependency injection doesn't make a whole lot of sense here if we're already being very prescriptive about the implementation

Reply:
For the first point - It actually does create an instance of the class [here](https://github.com/mongodb-js/compass/pull/5384/files#diff-b90c9f0d64dcf60e3a6638e888d2175e05cf9dff8f7ad68f4945619030e36ec6R193-R199) but at the same time I agree that `getStorage` would convey it much clearer that a new object is being returned.

> would it be possible to turn this into one?
I think something got missed here? What were you suggesting for this to be turned into?

Reply:
> `getStorage` would convey it much clearer that a new object is being returned

Well yeah, my point here was, it shouldn't matter whether it's a new object or not to callers :)

> What were you suggesting for this to be turned into?

Right now, the return type of `createStorage()`/`getStorage()` is `RecentQueryStorage`, which is a class, so from a TypeScript point of view it's not just a type, it's a *specific* implementation of that type. If we leave it like that, we cannot have alternate implementations of this service, which makes service injection a bit pointless – you're already stuck with a single, fixed implementation.

(Here's [a TS playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECSC2AHEBGaBvAUNaAzA9vgBQCUAXNAG74CWAJhgL7bSIBONlYALgKbR1e3MDRCkK1ek0zNMoSDATIATBhYFi5KrQbpmOdpx79Bw0eO1S9MzJgD0d6ABF8vCADsA5N2gB3fGwA1gA00ADmQkpiJNDAYO7QAFT47iAAnonQbEIArmwJ8dA07hDC7sD8+LhwSKiYxXxsuGAV0ABihAAKbPiU9LxsaFg4EdxRFlEoANw2cimlRbUoHfjdvf2DFCtrfYKD0AC8GOGRtaTH2dx5Ce68vjXIKKRT0IzMwPM+NLXK2z27AxQWy6-w2aCO6BOYzOMUhl2u0Fu9yiymer1kDmgAHUAoEIKFuAALfi8AAeiF4wD4DAaA2arRoMDJyBowBo3Hq7ka9P4AGFwFAsezCSs1DgNBZJHQZrJaU0Wvw-us9qphlDxlp+QohUSVjLbB8Sl8lkqAWxlMDVqCVYdjqMNRdcvlEXcHqg0W85kbFipTRsLe0QcqBqoIeqYY6rs6kW7USQXp6gA) that shows the problem as a concrete error, in case that helps clarify what I mean 🙂)

Very broadly speaking, we shouldn't rely on types from the "Compass"/"default" implementation in our provider.ts files, we should define interfaces here in the provider.ts file and then just make sure that the implementation implements that interface

Reply:
Ahh now it clicked to me! I definitely overlooked that aspect, thank you so much for pointing this out and for the playground also, definitely helped 🙌

Reply:
Addressed both in my recent commit. Now the compass specific implementation is separated from the interface itself for both pipeline storage and query storage. I renamed the compass specific implementation to be prefixed with `Compass` and the normal one is for the interface itself.